service/web/subscriberserver.go

```diff
@@ -34,20 +39,115 @@ func (s subscriberserver) CreateSubscription(ctx context.Context, req *metrov1.S
 
 // Acknowledge a message
 func (s subscriberserver) Acknowledge(ctx context.Context, req *metrov1.AcknowledgeRequest) (*emptypb.Empty, error) {
-	// TODO: Implement
+	logger.Ctx(ctx).Infow("received request to ack messages")
 	return new(emptypb.Empty), nil
 }
 
 // Pull messages
 func (s subscriberserver) Pull(ctx context.Context, req *metrov1.PullRequest) (*metrov1.PullResponse, error) {
 	logger.Ctx(ctx).Infow("received request to pull messages")
-	// TODO: Implement
+	// non streaming pull not to be supported
+	/*
+		res, err := s.subscriberCore.Pull(ctx, &subscriber.PullRequest{req.Subscription, 0, 0}, 2, xid.New().String()) // TODO: fix
+		if err != nil {
+			logger.Ctx(ctx).Errorw("pull response errored", "msg", err.Error())
+			return nil, merror.ToGRPCError(err)
+		}
+		return &metrov1.PullResponse{ReceivedMessages: res.ReceivedMessages}, nil
+
+	*/
 	return &metrov1.PullResponse{}, nil
 }
 
 // StreamingPull ...
 func (s subscriberserver) StreamingPull(server metrov1.Subscriber_StreamingPullServer) error {
-	// TODO: Implement
+	var pullStream *pullStream
+	var req *metrov1.StreamingPullRequest
+	var timeout *time.Ticker
+	var err error
+	timeout = time.NewTicker(5 * time.Second) // init with some sane value
+	streamResponseChan := make(chan metrov1.PullResponse)
+	reqChan := make(chan *metrov1.StreamingPullRequest)
+	errChan := make(chan error)
+	ctx := server.Context()
+	for {
+		// receive request in a goroutine, to timeout on stream ack deadline seconds
+		go receive(server, reqChan, errChan)
+		logger.Ctx(ctx).Info("loop")
+		select {
+		case <-ctx.Done():
+			logger.Ctx(ctx).Infow("returning after context is done")
+			if pullStream != nil {
+				pullStream.stop()
+			}
+			return ctx.Err()
+		// stream ack deadline
+		case <-timeout.C:
+			if pullStream != nil {
+				pullStream.stop()
+			}
+			return fmt.Errorf("stream ack deadline seconds crossed")
+		case err := <-errChan:
+			if pullStream != nil {
+				pullStream.stop()
+			}
+			if err == io.EOF {
+				// return will close stream from server side
+				logger.Ctx(ctx).Info("EOF received from client")
+				return nil
+			}
+			if err != nil {
+				logger.Ctx(ctx).Infow("error received from client", "msg", err.Error())
+				return nil
+			}
+		case req = <-reqChan:
+			// reset stream ack deadline seconds
+			timeout.Stop()
+			timeout = time.NewTicker(time.Duration(req.StreamAckDeadlineSeconds) * time.Second)
+			// ping request received with empty payload
+			if req.Subscription == "" {
```

Comment:
- should we add some validations on the pull request since `Subscription` is sent only when establishing a stream, thereafter only `AckIds` or `ModAckIds` will be sent. And an empty `Subscription` is allowed only when non-empty `AckIds` and `ModAckIds` are in the payload.

- Maybe another validation when an `ackId` or `modAck` Id comes in as the first request.

Reply:
Ack